addressmanager/addressmanager.go

```diff
@@ -0,0 +1,1363 @@
+
+// HostToNetAddress returns a netaddress given a host address. If
+// the host is not an IP address it will be resolved.
+func (am *AddressManager) HostToNetAddress(host string, port uint16, services wire.ServiceFlag) (*wire.NetAddress, error) {
+	ip := net.ParseIP(host)
+	if ip == nil {
+		ips, err := am.lookupFunc(host)
+		if err != nil {
+			return nil, err
+		}
+		if len(ips) == 0 {
+			return nil, errors.Errorf("no addresses found for %s", host)
+		}
+		ip = ips[0]
+	}
+
+	return wire.NewNetAddressIPPort(ip, port, services), nil
+}
+
+// NetAddressKey returns a key in the form of ip:port for IPv4 addresses
+// or [ip]:port for IPv6 addresses for use as keys in maps.
+func NetAddressKey(netAddress *wire.NetAddress) AddressKey {
+	port := strconv.FormatUint(uint64(netAddress.Port), 10)
+
+	return AddressKey(net.JoinHostPort(netAddress.IP.String(), port))
+}
+
+// GetAddress returns a single address that should be routable. It picks a
+// random one from the possible addresses with preference given to ones that
+// have not been used recently and should not pick 'close' addresses
+// consecutively.
+func (am *AddressManager) GetAddress() *KnownAddress {
+	// Protect concurrent access.
+	am.mutex.Lock()
+	defer am.mutex.Unlock()
+
+	triedAddressBucketArray := am.triedAddressBucketArray(am.localSubnetworkID)
+	triedAddressCount := am.triedAddressCount(am.localSubnetworkID)
+	newAddressBucketArray := am.newAddressBucketArray(am.localSubnetworkID)
+	newAddressCount := am.newAddressCount(am.localSubnetworkID)
+	knownAddress := am.getAddress(triedAddressBucketArray, triedAddressCount, newAddressBucketArray, newAddressCount)
+
+	return knownAddress
+
+}
+
+// getAddress returns a single address that should be routable.
+// See GetAddress for further details.
+func (am *AddressManager) getAddress(triedAddressBucketArray *triedAddressBucketArray, triedAddressCount int,
+	newAddressBucketArray *newAddressBucketArray, newAddressCount int) *KnownAddress {
+
+	// Use a 50% chance for choosing between tried and new addresses.
+	var bucketArray addressBucketArray
+	if triedAddressCount > 0 && (newAddressCount == 0 || am.random.Intn(2) == 0) {
+		bucketArray = triedAddressBucketArray
+	} else if newAddressCount > 0 {
+		bucketArray = newAddressBucketArray
+	} else {
+		// There aren't any addresses in any of the buckets
+		return nil
+	}
+
+	// Pick a random bucket
+	randomBucket := bucketArray.randomBucket(am.random)
+
+	// Get the sum of all chances
+	totalChance := float64(0)
+	for _, knownAddress := range randomBucket {
+		totalChance += knownAddress.chance()
+	}
+
+	// Pick a random address weighted by chance
+	randomValue := am.random.Float64()
+	accumulatedChance := float64(0)
+	for _, knownAddress := range randomBucket {
+		normalizedChance := knownAddress.chance() / totalChance
+		accumulatedChance += normalizedChance
+		if randomValue < accumulatedChance {
+			return knownAddress
+		}
+	}
+
+	panic("randomValue is exactly 1, which cannot happen")
```

Comment:
At the same time it could be also > 1.
I'd rephrase the error to "randomValue >= 1, which cannot happen".